File: backend/app/models/product.py
from enum import Enum
from sqlalchemy import Column, Float, ForeignKey, Integer, String, ARRAY, Date
from sqlalchemy.orm import relationship
from app.db.base_class import Base
from app.models.address import Address
from .user import User


class ProductCategory(str, Enum):
    FRUITS = "FRUITS"
    VEGETABLES = "VEGETABLES"


class Product(Base):
    image = Column(String(150), nullable=False)
    product_varieties = Column(String(100))
    name = Column(String(100), nullable=False)
    category: ProductCategory = Column(String(20), nullable=False)
    description = Column(String(1000))
    expected_available_date = Column(Date(), nullable=False)
    measurement_unit = Column(String(30), nullable=False)
    price = Column(Float(), nullable=False)
    available_quantity = Column(Float(), nullable=False)
    owner_id = Column(ForeignKey(f"{User.__tablename__}.id"), nullable=False)
    address_id = Column(ForeignKey(f"{Address.__tablename__}.id"), nullable=False)
    address = relationship(Address.__tablename__)
    owner = relationship(User.__tablename__)

    @property
    def image_url(self):
        return f"/{self.owner_id}/products/{self.id}/{self.image}"

    @property
    def varieties(self):
        if self.product_varieties:
            return self.product_varieties.split(",")
        return None

File: market/lib/data/models/product/product.dart
import 'dart:io';

import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:market/data/models/address/address.dart';
import 'package:market/data/models/user/user.dart';
import 'package:market/data/services/rest/client.dart';

part 'product.g.dart';
part 'product.freezed.dart';

@Freezed(fromJson: true, copyWith: true)
class Product with _$Product {
  Product._();
  factory Product({
    required int id,
    @JsonKey(name: "image_url") required String image,
    String? description,
    required User owner,
    required String name,
    required Address address,
    @JsonKey(name: "expected_available_date") required DateTime availableDate,
    required double price,
    @JsonKey(name: "measurement_unit") required String measurementUnit,
  }) = _Product;
  
  String get imageUrl => MEDIA_URL + image;
  factory Product.fromJson(json) => _$$_ProductFromJson(json);
}

@Freezed(fromJson: true)
class ProductsPage with _$ProductsPage {
  factory ProductsPage({
    required int pages,
    required List<Product> products,
    @JsonKey(name: "current_page") required int currentPage,
  }) = _ProductsPage;
  factory ProductsPage.fromJson(json) => _$$_ProductsPageFromJson(json);
}

@Freezed(toJson: true)
class ProductCreate with _$ProductCreate {
  factory ProductCreate({
    @Default("") String description,
    @JsonKey(name: "owner_id") required String ownerId,
    @JsonKey(name: "expected_available_date") required DateTime availableDate,
    required double price,
    required String name,
    @JsonKey(name: "address_id") required int addressId,
    @JsonKey(name: "available_quantity") required double availableQuantity,
    @JsonKey(name: "measurement_unit") required String measurementUnit,
  }) = _ProductCreate;
}

@Freezed(toJson: true)
class ProductUpdate with _$ProductUpdate {
  factory ProductUpdate({
    String? description,
    @JsonKey(name: "expected_available_date") DateTime? availableDate,
    double? price,
    String? name,
    @JsonKey(name: "available_quantity") double? availableQuantity,
    @JsonKey(name: "measurement_unit") String? measurementUnit,
  }) = _ProductUpdate;
}
